lib/src/vector.dart

```diff
@@ -0,0 +1,419 @@
+import 'package:meta/meta.dart';
+
+/// The Mask
+const int mask = (1 << 5) - 1;
+
+/// 32
+const int trieSize = 32;
+
+/// [Vector] is a persistent bit-partitioned Vector trie based on Clojure
+/// and Scala implementations.
+abstract class Vector<T> implements Iterable<T> {
+  const Vector._();
+
+  /// Creates a new Vector from any interable.
+  ///
+  /// example:
+  ///   final xs = new Vector.fromIterable([1, 2, 3, 4]);
+  ///   print(xs);
+  ///   => [1, 2, 3, 4]
+  factory Vector.fromIterable(Iterable<T> xs) {
+    if (xs.isEmpty) {
+      return new _ImmutableVector.empty();
+    }
+    final length = xs.length;
+    // calculate the necessary depth of the tree by bit shifting by 2^5.
+    int shift = length;
+    int depth = 0;
+    while (shift > 0) {
+      depth++;
+      shift = shift >> 5;
+    }
+    final root = fromDepth/*<T>*/(depth);
+    var index = 0;
+    for (final x in xs) {
+      root.set(index, 5 * (depth - 1), x);
+      index++;
+    }
+    return new _ImmutableVector.fromRoot(length, depth, root);
+  }
+
+  /// Creates an empty Vector.
+  factory Vector.empty() => new _ImmutableVector.empty();
+
+  @override
+  T get first {
+    if (length == 0) {
+      throw new StateError('Empty Vector has no first value.');
+    }
+    return this[0];
+  }
+
+  @override
+  bool get isEmpty => length == 0;
+
+  @override
+  bool get isNotEmpty => length != 0;
+
+  @override
+  Iterator<T> get iterator => new _VectorIterator(this);
+
+  @override
+  T get last {
+    if (length == 0) {
+      throw new StateError('Vector is empty and has no last value.');
+    }
+    return this[length - 1];
+  }
+
+  @override
+  int get length;
+
+  @override
+  T get single {
+    if (length == 1) {
+      return this[0];
+    }
+    throw new StateError('Vector has length $length');
+  }
+
+  /// used to implemented [Iterable] methods.
+  @protected
+  Iterable<T> get _values;
+
+  /// Index operator works the same as a normal list.
+  ///
+  /// Throws a state error if the index is out of bounds.
+  T operator [](int index);
+
+  /// Returns a Vector without the last item
+  ///
+  /// Throws a state error if it is empty.
+  Vector<T> remove();
```

Comment:
maybe it should be removeLast, basically I need a way to shrink the Vector.  maybe just something like slice(lower, upper)?